Automattic/newspack-network
# feat(content-distribution): editor plugin for distribution

Implements an editor plugin for distributing a post.

<img width="350" alt="image" src="https://github.com/user-attachments/assets/ba9dc863-cc7a-4f8e-b5e3-0d5d09dcd373" />

My original intention was to strictly leverage `editPost()` and `savePost()` dispatches to update the distribution post meta and let our short-circuit hook handle validation. This did not work well for error handling, as there was no straightforward way to pass the error message to the editor. Because of that, I switched to an API endpoint.

### Selecting connections and distributing

![newspack sa ngrok io_wp-admin_post php_post=328 action=edit](https://github.com/user-attachments/assets/d4779762-aac3-4710-9158-1a8575254a7d)

![newspack sa ngrok io_wp-admin_post php_post=328 action=edit (1)](https://github.com/user-attachments/assets/32cfb5b5-b35e-410e-ac8a-bc1f4b5da9af)

![newspack sa ngrok io_wp-admin_post php_post=328 action=edit (2)](https://github.com/user-attachments/assets/3770e94f-3612-458e-9b98-69488a976f9f)

![newspack sa ngrok io_wp-admin_post php_post=328 action=edit](https://github.com/user-attachments/assets/719c5faf-c2f0-46dd-b711-d955e424d1fa)

![newspack sa ngrok io_wp-admin_post php_post=328 action=edit (3)](https://github.com/user-attachments/assets/4dcdb13e-c453-401b-b373-42e1abc8218d)

### Updating connections

![newspack sa ngrok io_wp-admin_post php_post=326 action=edit](https://github.com/user-attachments/assets/a7652dba-6116-412b-99fb-78984dc1930e)

![newspack sa ngrok io_wp-admin_post php_post=326 action=edit (1)](https://github.com/user-attachments/assets/328ed07f-6fa1-4374-a833-a1de5b97eee9)

![newspack sa ngrok io_wp-admin_post php_post=326 action=edit](https://github.com/user-attachments/assets/d3f0af64-a4d2-44a6-adbd-158f18474c26)

![newspack sa ngrok io_wp-admin_post php_post=326 action=edit (3)](https://github.com/user-attachments/assets/7968b91e-39ea-46e6-a64c-2cc3e0fa9664)

### Error message

![newspack sa ngrok io_wp-admin_post php_post=327 action=edit](https://github.com/user-attachments/assets/4be9b6c7-9083-46c7-b3d0-864b90c8f720)

### Distributing an unsaved post

To prevent distributing an outdated version of the post, if the post has unsaved changes we'll enforce a post save before distributing:

![newspack sa ngrok io_wp-admin_post php_post=352 action=edit](https://github.com/user-attachments/assets/da59e532-3edf-42ba-ab85-350bc47f773f)

### A lot of nodes

![newspack sa ngrok io_wp-admin_post php_post=352 action=edit (1)](https://github.com/user-attachments/assets/cb0f5e73-be21-47b7-9acf-54dd7cc4a99a)

## Testing

**This PR also inaugurates javascript assets to the plugin. Before anything, make sure to run `npm ci && npm run build`**

1. Make sure you have a network setup with a hub and 2 nodes and `define( 'NEWPACK_NETWORK_CONTENT_DISTRIBUTION', true );` on all instances config
2. Draft a new post in the hub and click the globe icon as referenced in the first image above
3. Confirm the nodes are available for distribution
4. Make changes to the post and confirm the button changes from "Distribute" to "Save & Distribute"
5. Save the post and confirm the button is back to "Distribute"
6. Select one node and click "Distribute"
7. Confirm you get the pending state as in the image above and the snackbar notice on success
8. Confirm the selected node is now checked and disabled
9. Click "Select all" and confirm the other node gets selected
10. Click Distribute and confirm you get the same successful result
11. Confirm on the nodes that the post is distributed
12. Edit the post in one of the nodes and confirm the editor plugin (globe icon) is not available

### Multiple nodes

1. Replace line 57 of `src/content-distribution/distribute/index.js` with:
```js
	const sites = [];
	for( let i = 0; i < 100; i++ ) {
		sites.push( `https://node${ i + 1}.local` );
	}
```
2. Edit a post and confirm the plugin sidebar renders with a scrollbar

### Error handling

1. Add the following to line 78 of `includes/content-distribution/class-outgoing-post.php`: 
```php
return new WP_Error( 'test_error', 'This is my error message. Try distributing again.' );
```
2. Attempt to distribute a post via the editor and confirm you get the error notice.
